---
title: "Enhancers_SSMs_overlaps"
output: html_document
date: "2024-03-11"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
library(tidyverse)
library(GenomicRanges)
library(reshape2)

SEED <- 4321
set.seed(SEED)

```

```{r}
source("/Users/ieo6983/Desktop/fragile_enhancer_clinical/utils/functions_genomics.R")

IN_FOLDER <- "/Users/ieo6983/Desktop/fragile_enhancer_clinical/"
OUT_FOLDER_DATA <- "/Users/ieo6983/Desktop/fragile_enhancer_clinical/results/ICGC/enhancers_SSMs_overlaps/data/"
OUT_FOLDER_PLOTS <- "/Users/ieo6983/Desktop/fragile_enhancer_clinical/results/ICGC/enhancers_SSMs_overlaps/plots/"

```

```{r}
variants_gianluca <- F
variants_AFs <- T

if(variants_gianluca == T){
  # Variants dataset from Gianluca
  SSMs <- read_tsv("/Users/ieo6983/Desktop/fragile_enhancer_clinical/data/genomics/pre_processed_ICGC/from_gianluca.ssm.pcawg.coord.tsv")
  SSMs$end <- SSMs$POS
  SSMs <- SSMs %>% relocate(., end, .after = POS)
  colnames(SSMs)[c(3, 4, 5)] <- c("seqnames", "start", "end")
  SSMs$icgc_mutation_id <- str_c(SSMs$Type, SSMs$seqnames, SSMs$start, SSMs$end, sep = "_")
  SSMs_gr <- makeGRangesFromDataFrame(SSMs, keep.extra.columns = T)
} else if(variants_AFs == T){
  # Variants dataset with AFs
  SSMs <- read_tsv("/Users/ieo6983/Desktop/fragile_enhancer_clinical/data/genomics/pre_processed_ICGC/simple_somatic_mutation.open.matching_calls.with_AFs.tsv")
  colnames(SSMs)[c(7,8)] <- c("start", "end")  
  SSMs$chromosome <- paste("chr", SSMs$chromosome, sep = "")
  
  # Add mutatation ID equal to hartwig mutation ID
  SSMs$ID <- str_sub(SSMs$chromosome, start = 4) %>% 
    str_c(., SSMs$end, sep = ":") %>% 
    str_c(., SSMs$reference_genome_allele, sep = "_") %>%
    str_c(., SSMs$mutated_to_allele, sep = "/")
   
  SSMs_gr <- makeGRangesFromDataFrame(SSMs, keep.extra.columns = T)
}


```


### Input files ###
```{r}
# SSMs
if(variants_gianluca == F & variants_AFs == F){
  SSMs <- read_tsv(paste(IN_FOLDER, "./data/genomics/pre_processed_ICGC/simple_somatic_mutation.open.matching_calls.tsv", sep=""))
  SSMs <- SSMs[, -c(9,14)]
  colnames(SSMs)[c(7,8)] <- c("start", "end") 
  SSMs$chromosome <- paste("chr", SSMs$chromosome, sep = "")
  SSMs_gr <- makeGRangesFromDataFrame(SSMs, keep.extra.columns = T)
}

# ALL enhancers: CtIP, GRHL
columns_names <- c("chrom", "start", "end", "cluster")
enh_ctip <- read_tsv(paste(IN_FOLDER, "data/functional_genomics/others/Cluster_CtIP_Enh_All.txt", sep=""), col_names = columns_names,
                     comment = "#")
enh_grhl <- read_tsv(paste(IN_FOLDER, "data/functional_genomics/others/Cluster_GRHL_Enh_All.txt", sep=""), col_names = columns_names, 
                     comment = "#")
enh_all <- list("CtIP" = enh_ctip, "GRHL" = enh_grhl)
# add summit & name
enh_all <- lapply(enh_all, function(df){
  df$summit <- df$end
  df$name <- str_c(df$chrom, df$summit, sep=":")
  return(df)})


# Putative enhancers: H3K27ac (no CtIP, GRHL, MRE11, self_overlaps)
put_enhancers <- read_tsv(paste(IN_FOLDER, "data/functional_genomics/others/PutativeEnhnacers_MCF10A_hg19.filtered_GRHL_CtIP_MRE11_self.tsv", sep=""))
colnames(put_enhancers)[1] <- "chrom" 
put_enhancers$summit <- put_enhancers$end


### 
WIN <- 3000
UNIT <- "kb"
MARKERS <- names(enh_all)

```




### Find enhancers / SSMs overlaps ###


## SSMs overlap over ALL enhancers
```{r}
input_variants <- SSMs
input_variants_gr <- SSMs_gr

stat_enhancers_enh <- matrix(nrow = 6000, ncol = length(MARKERS))
colnames(stat_enhancers_enh) <- MARKERS
stat_donors_enh <- matrix(nrow = length(unique(input_variants$icgc_donor_id)), ncol = length(MARKERS))
colnames(stat_donors_enh) <- MARKERS

for(marker in MARKERS){
  mut <- "SSMs"
  
  ### Pre/process input data ###
  # CtIP/GRHL enhancers
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'
  # Putative enhancers
  put_enh <- put_enhancers[sample(seq(1,dim(put_enhancers)[1]), size = dim(marker_enh)[1], replace = F), ] 
  
  ## Extend regions of 'win' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN
  put_enh$start <- put_enh$summit - WIN
  put_enh$end <- put_enh$summit + WIN
  
  # Generate random input sequences 
  ran_seqs <- generate_random_seqs(SEED, dim(marker_enh)[1], win = WIN)
  
  # Create GRanges objects 
  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)
  put_enh_gr <- makeGRangesFromDataFrame(put_enh, keep.extra.columns = T)
  ran_seqs_gr = makeGRangesFromDataFrame(ran_seqs, keep.extra.columns = T)
  
  
  
  ### Find mutations / enhancers overlaps ###
  # CtIP/GRHL enhancers - overlap with matching SSMs 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=SSMs_gr)
  SSMs_sbj <- data.frame(SSMs_gr[subjectHits(hits_obj_enh)])
  colnames(SSMs_sbj)[1:5] <- paste(colnames(SSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), SSMs_sbj) 
  ## Controls:
  # Putative enhancers
  hits_obj_put_enh <- findOverlaps(query = put_enh_gr, subject = SSMs_gr)
  SSMs_sbj <- data.frame(SSMs_gr[subjectHits(hits_obj_put_enh)])
  colnames(SSMs_sbj)[1:5] <- paste(colnames(SSMs_sbj)[1:5], "sbj", sep = "_")
  put_enh_SSMs <- cbind(data.frame(put_enh_gr[queryHits(hits_obj_put_enh)]), SSMs_sbj) 
  # Random sequences
  hits_obj_ran <- findOverlaps(query=ran_seqs_gr, subject=SSMs_gr)
  SSMs_sbj <- data.frame(SSMs_gr[subjectHits(hits_obj_ran)])
  colnames(SSMs_sbj)[1:5] <- paste(colnames(SSMs_sbj)[1:5], "sbj", sep = "_")
  ran_seqs_SSMs <- cbind(data.frame(ran_seqs_gr[queryHits(hits_obj_ran)], SSMs_sbj))
  
  
  
  ### SAVE stats - Enhancers only
  save_intermediate <- F
  if(save_intermediate == TRUE){
    
    # Save enh_SSMs objects - intermediate object useful for all the ubsequent analyses
    enh_SSMs <- enh_SSMs %>% relocate(., summit, .before = start) 
    write_tsv(enh_SSMs, 
              paste0(OUT_FOLDER_DATA, sprintf("Table_enh_SSMs_%s.all_overlaps.%s%s_WIN.with_AFs.tsv", marker, WIN/1000, UNIT)))
    
  }

  ### Compute stats for plots - all datasets
  classes = c(paste(marker,"enhancers",sep="_"), "put_enhancers", "ran_seqs")
  stat_enh <- compute_stat_enhancers(list(enh_SSMs, put_enh_SSMs, ran_seqs_SSMs), 
                                     list(marker_enh, put_enh, ran_seqs), classes = classes)
  stat_don <- compute_stat_donors(list(enh_SSMs, put_enh_SSMs, ran_seqs_SSMs), SSMs, classes = classes)
  
  # Add enhancers stats 
  stat_enhancers_enh[, marker] <- c(stat_enh[, classes[1]], 
                                    rep(NA, dim(stat_enhancers_enh)[1] - length(stat_enh[, classes[1]])))
  stat_donors_enh[, marker] <- stat_don[, classes[1]]
  
  
  ### Plot stats
  avg <- str_flatten(paste(names(apply(stat_enh, 2, mean)), ":", round(apply(stat_enh, 2, mean),2), sep = ""), 
                     collapse = "  ")
  perc <- apply(stat_enh, 2, function(x) round(sum(x != 0) / dim(marker_enh)[1] * 100))
  perc <- str_flatten(paste(names(perc), ":", perc, " %", sep=""), collapse = "  ")
  
  b1 <- stat_enh %>%
    pivot_longer(everything(), names_to = "cluster", values_to = "value") %>%
    ggplot(., aes(x = cluster, y = value, fill = cluster))+
    geom_boxplot()+
    labs(title = "Number of donors hit", 
         subtitle = paste(
           paste("Average number of donors that each enhancer hits: ", avg),
           paste("\nPercentage or enhancers mutated in at least 1 donor: ", perc)
         ),
         x = "", y = "number of donors")+
    theme_light()+
    theme(axis.text.x = element_blank(), 
          plot.subtitle = element_text(size = 10))+
    scale_y_continuous(limits = c(0,15))
  print(b1)
  #ggsave(paste(OUT_FOLDER_PLOTS, paste("Stat_enhancers", marker, "all_enhancers", mut, sep = "_"), ".png", sep = ""), plot = b1, device = "png", width = 10, height = 7)
  

  # P-values
  print("Wilcox - n_donors")
  print(wilcox.test(stat_enh[,1], stat_enh[,2])[["p.value"]])
  print(wilcox.test(stat_enh[,1], stat_enh[,3])[["p.value"]])
  
  print("T-test - n_donors")
  print(t.test(stat_enh[,1], stat_enh[,2])[["p.value"]])
  print(t.test(stat_enh[,1], stat_enh[,3])[["p.value"]])
  
  avg <- str_flatten(paste(names(apply(stat_don, 2, mean)), ":", round(apply(stat_don, 2, mean),2), sep = ""), 
                     collapse = "  ")
  perc <- apply(stat_don, 2, function(x) round(sum(x != 0) / length(unique(SSMs$icgc_donor_id)) * 100))
  perc <- str_flatten(paste(names(perc), ":", perc, " %", sep=""), collapse = "  ")
  b2 <- stat_don %>%
    pivot_longer(everything(), names_to = "cluster", values_to = "value") %>%
    ggplot(., aes(x = cluster, y = log10(value), fill = cluster))+
    geom_boxplot()+
    labs(title = paste("Number of enahncers hit in each donor", sep=""), 
         subtitle = paste(
           paste("Average number of enhancers hit in a donor: ", avg), 
           paste("\nPercentage of donors with at least one enhancer mutated : ", perc)),
         x = "", y = "log10(number of enhancers hit)")+
    theme_light()+
    theme(axis.text.x = element_blank(), plot.subtitle = element_text(size = 10))
  print(b2)
  #ggsave(paste(OUT_FOLDER_PLOTS, paste("Stat_donors", marker, "all_enhancers", mut, sep = "_"), ".png", sep = ""), plot = b1, device = "png", width = 10, height = 7)
  
  # P-values
  print(wilcox.test(stat_don[,1], stat_don[,2])[["p.value"]])
  print(wilcox.test(stat_don[,1], stat_don[,3])[["p.value"]])
}

```



## SSMs overlap over High-low enhancers
```{r}
enh_SSMs_ctip <- read_tsv(paste(OUT_FOLDER_DATA, "Table_enh_SSMs_CtIP.all_overlaps.tsv", sep=""))
enh_SSMs_grhl <- read_tsv(paste(OUT_FOLDER_DATA, sprintf("Table_enh_SSMs_GRHL.all_overlaps.%s%s_WIN.tsv", WIN,UNIT), sep=""))
enh_SSMs <- list("CtIP" = enh_SSMs_ctip, "GRHL" = enh_SSMs_grhl)


```

```{r}
label <- "high_low_enhancers"

# Splitting based on clustering 
clusters_high_ctip <- c("CtIP_cluster_2_Enh", "CtIP_cluster_3_Enh", "CtIP_cluster_5_Enh")
clusters_high_grhl <- c("GRHL_cluster_1_Enh", "GRHL_cluster_2_Enh", "GRHL_cluster_4_Enh")
clusters_high <- list("CtIP" = clusters_high_ctip, "GRHL" = clusters_high_grhl)

# High-low stats for CtIP
stat_enhancers_enh_ctip <- matrix(nrow = 6000, ncol = 2)
colnames(stat_enhancers_enh_ctip) <- c("high", "low")
stat_donors_enh_ctip <- matrix(nrow = length(unique(input_variants$icgc_donor_id)), ncol = 2)
colnames(stat_donors_enh_ctip) <- c("high", "low")
# High-low stats for GRHL
stat_enhancers_enh_grhl <- matrix(nrow = 6000, ncol = 2)
colnames(stat_enhancers_enh_grhl) <- c("high", "low")
stat_donors_enh_grhl <- matrix(nrow = length(unique(input_variants$icgc_donor_id)), ncol = 2)
colnames(stat_donors_enh_grhl) <- c("high", "low")

stat_enhancers_enh <- list("Ctip" = stat_enhancers_enh_ctip, "GRHL" = stat_enhancers_enh_grhl)



for(marker in MARKERS){
  mut <- "SSMs"
  marker_enh_high <- enh_all[[marker]] %>% filter(cluster %in% clusters_high[[marker]])
  marker_enh_low <- enh_all[[marker]] %>% filter(!cluster %in% clusters_high[[marker]])
    
  overlaps <- enh_SSMs[[marker]]
  overlaps_high <- overlaps %>% filter(cluster %in% clusters_high[[marker]])
  overlaps_low <- overlaps %>% filter(!cluster %in% clusters_high[[marker]])
  
  
  
  ## STATS & PLOTS ##
  classes = c(paste(marker,"enhancers", "high",sep="_"), paste(marker,"enhancers", "low",sep="_"))
  stat_enh <- compute_stat_enhancers(list(overlaps_high, overlaps_low), 
                                     list(marker_enh_high, marker_enh_low), classes = classes)
  stat_enh <- na.omit(stat_enh)
  stat_don <- compute_stat_donors(list(overlaps_high, overlaps_low), SSMs, classes = classes)
  stat_don <- na.omit(stat_don)
  
  
  ### Plot stats enh
  avg <- str_flatten(paste(names(apply(stat_enh, 2, mean)), ":", round(apply(stat_enh, 2, mean),2), sep = ""), 
                     collapse = "  ")
  #perc <- apply(stat_enh, 2, function(x) round(sum(x != 0) / dim(marker_enh)[1] * 100))
  perc <- apply(stat_enh, 2, function(x) round(sum(x != 0)))
  perc <- round(perc / c(dim(marker_enh_high)[1], dim(marker_enh_low)[1]) * 100,2)
  perc <- str_flatten(paste(names(perc), ":", perc, " %", sep=""), collapse = "  ")
  
  b1 <- stat_enh %>%
    pivot_longer(everything(), names_to = "cluster", values_to = "value") %>%
    ggplot(., aes(x = cluster, y = value, fill = cluster))+
    geom_boxplot()+
    labs(title = "Number of donors hit", 
         subtitle = paste(
           paste("Average number of donors that each enhancer hits: ", avg),
           paste("\nPercentage or enhancers mutated in at least 1 donor: ", perc)
         ),
         x = "", y = "number of donors")+
    theme_light()+
    theme(axis.text.x = element_blank(), 
          plot.subtitle = element_text(size = 10))+
    scale_y_continuous(limits = c(0,15))
  print(b1)

  
  # Plot stats don
  avg <- str_flatten(paste(names(apply(stat_don, 2, mean)), ":", round(apply(stat_don, 2, mean),2), sep = ""), 
                     collapse = "  ")
  perc <- apply(stat_don, 2, function(x) round(sum(x != 0) / length(unique(SSMs$icgc_donor_id)) * 100))
  perc <- str_flatten(paste(names(perc), ":", perc, " %", sep=""), collapse = "  ")
  
  b2 <- stat_don %>%
    pivot_longer(everything(), names_to = "cluster", values_to = "value") %>%
    ggplot(., aes(x = cluster, y = log10(value), fill = cluster))+
    geom_boxplot()+
    labs(title = paste("Number of enahncers hit in each donor", sep=""), 
         subtitle = paste(
           paste("Average number of enhancers hit in a donor: ", avg), 
           paste("\nPercentage of donors with at least one enhancer mutated : ", perc)),
         x = "", y = "log10(number of enhancers hit)")+
    theme_light()+
    theme(axis.text.x = element_blank(), plot.subtitle = element_text(size = 10))
  print(b2)


}

    

```

