---
title: "SNVs_distribution"
output: html_document
date: "2024-02-22"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tidyverse)
library(GenomicRanges)
library(reshape2)
library(hrbrthemes) # fro theme_ipsum()
library(moments) # to compute distribution skewness()

SEED <- 4321
set.seed(SEED)

```

```{r}
IN_FOLDER <- "/Users/ieo6983/Desktop/fragile_enhancer_clinical/data/"
OUT_FOLDER <- "/Users/ieo6983/Desktop/fragile_enhancer_clinical/results/ICGC/mutations_distribution_over_enhancers/"

```


### INPUT FILES
```{r}
# SSMs 
SSMs <- read_tsv(paste(IN_FOLDER, "./genomics/pre_processed_ICGC/simple_somatic_mutation.open.matching_calls.tsv", sep=""))
SSMs <- SSMs[, -c(9,14)]
colnames(SSMs)[c(7,8)] <- c("start", "end") 
SSMs$chromosome <- paste("chr", SSMs$chromosome, sep = "")
## Only single-base substitution
SSMs_snvs <- SSMs[SSMs$mutation_type == "single base substitution", ] # 4'608'346 (over ~5M)
SSMs_gr <- makeGRangesFromDataFrame(SSMs_snvs, keep.extra.columns = T)


# ALL enhancers: CtIP, GRHL
columns_names <- c("chrom", "start", "end", "cluster")
enh_ctip <- read_tsv(paste(IN_FOLDER, "functional_genomics/others/Cluster_CtIP_Enh_All.txt", sep=""), col_names = columns_names,
                     comment = "#")
enh_grhl <- read_tsv(paste(IN_FOLDER, "functional_genomics/others/Cluster_GRHL_Enh_All.txt", sep=""), col_names = columns_names, 
                     comment = "#")
enh_all <- list("CtIP" = enh_ctip, "GRHL" = enh_grhl)
# add summit & name
enh_all <- lapply(enh_all, function(df){
  df$summit <- df$end
  df$name <- str_c(df$chrom, df$summit, sep=":")
  return(df)})


# Putative enhancers: H3K27ac (no CtIP, GRHL, MRE11, self_overlaps)
put_enhancers <- read_tsv(paste(IN_FOLDER, "functional_genomics/others/PutativeEnhnacers_MCF10A_hg19.filtered_GRHL_CtIP_MRE11_self.tsv", sep=""))
colnames(put_enhancers)[1] <- "chrom" 
put_enhancers$summit <- put_enhancers$end


```



### DISTRIBUTION OF SNVs IN ENHANCER REGIONS ###
```{r}
label <- "all_enhancers_and_k27ac_control"
WIN <- 3000

for(marker in names(enh_all)){

  ### Pre/process input data
  # CtIP/GRHL enhancers
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'
  # Putative enhancers
  put_enh <- put_enhancers[sample(seq(1,dim(put_enhancers)[1]), size = dim(marker_enh)[1], replace = F), ]
  print(paste("Number of putative enhancers sampled: ", dim(put_enh)[1]))

  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN
  put_enh$start <- put_enh$summit - WIN
  put_enh$end <- put_enh$summit + WIN
  
  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)
  put_enh_gr <- makeGRangesFromDataFrame(put_enh, keep.extra.columns = T)

  
  ### Find SNVs / Enhancers overlaps 
  # All CtIP & GRHL enhancers - overlap with matching SSMs 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=SSMs_gr)
  SSMs_sbj <- data.frame(SSMs_gr[subjectHits(hits_obj_enh)])
  colnames(SSMs_sbj)[1:5] <- paste(colnames(SSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), SSMs_sbj) 

  hits_obj_put_enh <- findOverlaps(query = put_enh_gr, subject = SSMs_gr)
  SSMs_sbj <- data.frame(SSMs_gr[subjectHits(hits_obj_put_enh)])
  colnames(SSMs_sbj)[1:5] <- paste(colnames(SSMs_sbj)[1:5], "sbj", sep = "_")
  put_enh_SSMs <- cbind(data.frame(put_enh_gr[queryHits(hits_obj_put_enh)]), SSMs_sbj) 

  
  ### Compute dist of SNVs from enhancer summit 
  dist_enh <- enh_SSMs$start_sbj - enh_SSMs$summit
  dist_put_enh <- put_enh_SSMs$start_sbj - put_enh_SSMs$summit
  max_lenght <- max(length(dist_enh), length(dist_put_enh))
  dist_all <- data.frame(dist_enh = c(dist_enh, rep(NA, max_lenght - length(dist_enh))), 
                         dist_put_enh = c(dist_put_enh, rep(NA, max_lenght - length(dist_put_enh))))
    
  
  ### Plot binned distributions
  nbins <- c(1000, 100)
  ifelse(marker == "CtIP", color <- "lightblue", color <- "pink")
  
  for(i in 1:2){
    p <- dist_all %>% ggplot(.)+
      geom_histogram(aes(x = dist_enh), fill = color, alpha = 0.7, bins = nbins[i])+
      geom_text(data = data.frame(), aes(x=Inf, y=Inf, label= marker), color=color, vjust = 1.5, hjust = 2, 
                size = 7, fontface = "bold")+
      geom_histogram(aes(x = dist_put_enh, y = -..count..), fill = "gray", alpha = 0.7, bins = nbins[i])+
      geom_text(data=data.frame(), aes(x=Inf, y=-Inf, label= "H3K27ac_CTRL"), color="gray", vjust = -1, hjust = 1.1, 
                size = 7, fontface = "bold")+
      theme_light()+
      scale_x_continuous(breaks = seq(-3000, 3000, 500))+
      labs(title = paste("Histogram of distances from peak summit - ", marker, " (Binsize = ", WIN*2/nbins, "bp)", sep =""), 
         x = "distance from summit", y = "")
    print(p)
    
    ### Save Plot
    #ggsave(paste(paste(OUT_FOLDER, paste("SNVs_distribution", marker, label, "binsize", WIN*2/nbins[i], "bp", sep="_"), sep=""), ".png", sep=""), plot = p, device = "png", width = 18, height = 7)
    
  
  }
  
}
  
```


## SSMs distribution based on clusters ##
```{r}
save_plots <- FALSE

label <- "clustered_enhancers"
dist_clustered_hist <- list("CtIP" = setNames(replicate(length(unique(enh_all[["CtIP"]]$cluster)), list()),
                                               unique(enh_all[["CtIP"]]$cluster)), 
                             "GRHL" = setNames(replicate(length(unique(enh_all[["GRHL"]]$cluster)), list()),
                                               unique(enh_all[["GRHL"]]$cluster)))

for(marker in names(enh_all)){

  ### Pre/process input data
  # CtIP/GRHL enhancers
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'
  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN
  
  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)

  
  ### Find SNVs / Enhancers overlaps 
  # All CtIP & GRHL enhancers - overlap with matching SSMs 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=SSMs_gr)
  SSMs_sbj <- data.frame(SSMs_gr[subjectHits(hits_obj_enh)])
  colnames(SSMs_sbj)[1:5] <- paste(colnames(SSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), SSMs_sbj) 

  for(cluster in unique(marker_enh$cluster)){
    enh_SSMs_clust <- enh_SSMs[enh_SSMs$cluster == cluster, ]

    ### Compute dist of SNVs from enhancer summit 
    dist_enh <- enh_SSMs_clust$start_sbj - enh_SSMs_clust$summit
    max_lenght <- max(length(dist_enh), length(dist_put_enh))
    dist_all <- data.frame(dist_enh = c(dist_enh, rep(NA, max_lenght - length(dist_enh))))
    
    ### Plot binned distributions
    nbins <- 20
    ifelse(marker == "CtIP", color <- "lightblue", color <- "pink")
    title <- paste("Histogram of distances from peak summit - ", marker, " (Nbins = ", nbins, "bp)", sep ="")
    
    p <- dist_all %>% ggplot(.)+
      geom_histogram(aes(x = dist_enh), fill = color, alpha = 0.7, bins = nbins)+
      geom_text(data = data.frame(), aes(x=Inf, y=Inf, label= marker), color=color, vjust = 1.5, hjust = 2, 
                size = 7, fontface = "bold")+
      theme_light()+
      scale_x_continuous(breaks = seq(-3000, 3000, 500))+
      labs(title = NULL, 
           subtitle = cluster,
         x = "distance from summit", y = "")
    dist_clustered_hist[[marker]][[cluster]] <- p
    print(p)
    
  }
  
  # Save grid of plots 
  if(save_plots == T){
    title <- cowplot::ggdraw() + cowplot::draw_label(title, fontface = "bold")
    cowplot::plot_grid(title, plotlist = dist_clustered_hist[[marker]], ncol = 1) %>%
    ggsave(., filename = paste(paste(OUT_FOLDER, 
                       paste("SNVs_distribution", marker, label, "NBINS", nbins, "bp", sep="_"), sep=""), ".png", sep=""), 
        device = "png", width = 7, height = 18)

        
  }
}

```


# Density plot
## SSMs distribution based on clusters ##
```{r}
save_hist <- FALSE
save_density <- FALSE

label <- "clustered_enhancers"
dist_clustered_hist <- list("CtIP" = setNames(replicate(length(unique(enh_all[["CtIP"]]$cluster)), list()),
                                               unique(enh_all[["CtIP"]]$cluster)), 
                             "GRHL" = setNames(replicate(length(unique(enh_all[["GRHL"]]$cluster)), list()),
                                               unique(enh_all[["GRHL"]]$cluster)))
dist_clustered_density <- list("CtIP" = setNames(replicate(length(unique(enh_all[["CtIP"]]$cluster)), list()),
                                               unique(enh_all[["CtIP"]]$cluster)), 
                             "GRHL" = setNames(replicate(length(unique(enh_all[["GRHL"]]$cluster)), list()),
                                               unique(enh_all[["GRHL"]]$cluster)))

# Save all distances to be able to plot them together
all_distances <- list("CtIP" = NA, "GRHL" = NA)

for(marker in names(enh_all)){

  ### Pre/process input data
  # CtIP/GRHL enhancers
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'
  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN
  
  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)

  
  ### Find SNVs / Enhancers overlaps 
  # All CtIP & GRHL enhancers - overlap with matching SSMs 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=SSMs_gr)
  SSMs_sbj <- data.frame(SSMs_gr[subjectHits(hits_obj_enh)])
  colnames(SSMs_sbj)[1:5] <- paste(colnames(SSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), SSMs_sbj) 

  # Save all distances to be able to plot them together
  all_distances_marker <- data.frame(matrix(nrow = max(table(enh_SSMs$cluster)), ncol = length(unique(enh_SSMs$cluster))))
  colnames(all_distances_marker) <- unique(enh_SSMs$cluster)
  for(cluster in unique(marker_enh$cluster)){
    enh_SSMs_clust <- enh_SSMs[enh_SSMs$cluster == cluster, ]

    ### Compute dist of SNVs from enhancer summit 
    dist_enh <- enh_SSMs_clust$start_sbj - enh_SSMs_clust$summit
    max_lenght <- max(length(dist_enh))
    dist_all <- data.frame(dist_enh = c(dist_enh, rep(NA, max_lenght - length(dist_enh))))
    
    all_distances_marker[, cluster] <- c(dist_all[, "dist_enh"], rep(NA, dim(all_distances_marker)[1] - 
                                                                       length(dist_all[, "dist_enh"])))
    
    ### Plot binned distributions
    nbins <- c(100)
    ifelse(marker == "CtIP", color <- "lightblue", color <- "pink")
    title <- paste("Histogram of distances from peak summit - ", marker, " (Binsize = ", WIN*2/nbins, "bp)", sep ="")
    
    p <- dist_all %>% ggplot(.)+
      geom_histogram(aes(x = dist_enh), fill = color, alpha = 0.7, bins = nbins[1])+
      geom_text(data = data.frame(), aes(x=Inf, y=Inf, label= marker), color=color, vjust = 1.5, hjust = 2, 
                size = 7, fontface = "bold")+
      theme_light()+
      scale_x_continuous(breaks = seq(-3000, 3000, 500))+
      labs(title = NULL, 
           subtitle = cluster,
         x = "distance from summit", y = "")
    dist_clustered_hist[[marker]][[cluster]] <- p
    #print(p)
    
    title <- paste("Density of distances from peak summit - ", marker, sep ="")
    bw = 100
    dens_p <- dist_all %>% ggplot(.)+
      geom_density(aes(x = dist_enh), fill = color, color = color, alpha = 0.7, bw = bw)+
      geom_text(data = data.frame(), aes(x=Inf, y=Inf, label= marker), color=color, vjust = 1.5, hjust = 2, 
                size = 7, fontface = "bold")+
      theme_light()+
      scale_x_continuous(breaks = seq(-3000, 3000, 500))+
      labs(title = NULL, 
           subtitle = cluster,
         x = "distance from summit", y = "")
    dist_clustered_density[[marker]][[cluster]] <- dens_p
    
    print(dens_p)
    
  }
  
  all_distances[[marker]] <- all_distances_marker
  
  # Save grid of plots 
  if(save_hist == T){
    title <- cowplot::ggdraw() + cowplot::draw_label(title, fontface = "bold")
    cowplot::plot_grid(title, plotlist = dist_clustered_hist[[marker]], ncol = 1) %>%
    ggsave(., filename = paste(paste(OUT_FOLDER, 
                       paste("SNVs_distribution", marker, label, "binsize", WIN*2/nbins[1], "bp", sep="_"), sep=""), ".png", sep=""), 
        device = "png", width = 7, height = 20)
  }
  
  if(save_density == T){
    title <- cowplot::ggdraw() + cowplot::draw_label(title, fontface = "bold")
    cowplot::plot_grid(title, plotlist = dist_clustered_density[[marker]], ncol = 1) %>%
    ggsave(., filename = paste(paste(OUT_FOLDER, 
                       paste("SNVs_distribution", marker, label, "density_bandwidth", bw, "WIN", WIN, sep="_"), sep=""), ".png", sep=""), 
        device = "png", width = 7, height = 20)
  }
  
}


```

adjust = A multiplicate bandwidth adjustment. Adjust the bandwidth while still using the a bandwidth estimator.
bw = manually set the bandwidth.



_High-low clusters_
# Option 1: High = primi 3; Low = ultimi 4
clust_high_ctip <- c("CtIP_cluster_2_Enh", "CtIP_cluster_3_Enh", "CtIP_cluster_5_Enh")
clust_high_grhl <- c("GRHL_cluster_1_Enh", "GRHL_cluster_2_Enh", "GRHL_cluster_4_Enh")
clust_low_ctip <- c("CtIP_cluster_6.2_Enh", "CtIP_cluster_6.3_Enh", "CtIP_cluster_6.1_Enh", "CtIP_cluster_6.0")
clust_low_grhl <- c("GRHL_cluster_5.2_Enh", "GRHL_cluster_5.3_Enh", "GRHL_cluster_5.1_Enh", "GRHL_cluster_5.0")
# Option 2: High = primi 4; Low = ultimi 3
clust_high_ctip <- c("CtIP_cluster_2_Enh", "CtIP_cluster_3_Enh", "CtIP_cluster_5_Enh", "CtIP_cluster_6.2_Enh")
clust_high_grhl <- c("GRHL_cluster_1_Enh", "GRHL_cluster_2_Enh", "GRHL_cluster_4_Enh", "GRHL_cluster_5.2_Enh")
clust_low_ctip <- c("CtIP_cluster_6.3_Enh", "CtIP_cluster_6.1_Enh", "CtIP_cluster_6.0")
clust_low_grhl <- c("GRHL_cluster_5.3_Enh", "GRHL_cluster_5.1_Enh", "GRHL_cluster_5.0")
# Option 3: High = primi 5; Low = utlimi 2
clust_high_ctip <- c("CtIP_cluster_2_Enh", "CtIP_cluster_3_Enh", "CtIP_cluster_5_Enh", "CtIP_cluster_6.2_Enh", "CtIP_cluster_6.3_Enh")
clust_high_grhl <- c("GRHL_cluster_1_Enh", "GRHL_cluster_2_Enh", "GRHL_cluster_4_Enh", "GRHL_cluster_5.2_Enh", "GRHL_cluster_5.2_Enh")
clust_low_ctip <- c("CtIP_cluster_6.1_Enh", "CtIP_cluster_6.0")
clust_low_grhl <- c("GRHL_cluster_5.1_Enh", "GRHL_cluster_5.0")
```{r}
# Option 1: High = primi 3; Low = ultimi 4
clust_high_ctip <- c("CtIP_cluster_2_Enh", "CtIP_cluster_3_Enh", "CtIP_cluster_5_Enh")
clust_high_grhl <- c("GRHL_cluster_1_Enh", "GRHL_cluster_2_Enh", "GRHL_cluster_4_Enh")
clust_low_ctip <- c("CtIP_cluster_6.2_Enh", "CtIP_cluster_6.3_Enh", "CtIP_cluster_6.1_Enh", "CtIP_cluster_6.0")
clust_low_grhl <- c("GRHL_cluster_5.2_Enh", "GRHL_cluster_5.3_Enh", "GRHL_cluster_5.1_Enh", "GRHL_cluster_5.0")

clust_high <- list("CtIP" = clust_high_ctip, "GRHL" = clust_high_grhl)
clust_low <- list("CtIP" = clust_low_ctip, "GRHL" = clust_low_grhl)
clust_all <- list("high" = clust_high, "low" = clust_low)

```

```{r}
label <- "high_low_clusters"

for(marker in names(all_distances)){
  df <- all_distances[[marker]] 
  p <-  df %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(data=., aes(x=distances, group=cluster, color = cluster)) +
    geom_density(bw = 200, alpha = 0.1) +
    theme_light()
  print(p)
}

# Grouping distances based on high-low clusters
bw <- 100
for(marker in names(all_distances)){
  df <- all_distances[[marker]]
  high <- clust_all[["high"]][[marker]]
  low <- clust_all[["low"]][[marker]]
  
  dist_high <- na.omit(unlist(df[, colnames(df) %in% high, ], use.names = F))
  dist_low <-  na.omit(unlist(df[, colnames(df) %in% low, ], use.names = F))
  max_dist_length <- max(length(dist_high), length(dist_low))
  
  # How many SNVs fall in high vs. low enhancers?
  print(sprintf("Number of SNVs falling in %s enhancer region - WIN: %d", marker, WIN))
  print(sprintf("High: %d", length(dist_high)))
  print(sprintf("Low: %d", length(dist_low)))
  
  df_grouped <- data.frame("high" = c(dist_high, rep(NA, max_dist_length - length(dist_high))), 
                           "low" = c(dist_low, rep(NA, max_dist_length - length(dist_low))))
  

  # Density plot
  p1 <- df_grouped %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(., aes(x = distances, color = cluster, fill = cluster))+
  geom_density(bw = bw, alpha = 0.2)+
  theme_light()
  print(p1)
  
  p2 <- df_grouped %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(., aes(x = distances, group = cluster, color = cluster))+
  geom_density(bw = bw, alpha = 0.2, size = 1.5)+
  theme_light()
  print(p2)
  
  # Normalized density plot
  p1 <- df_grouped %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(., aes(x = distances, group = cluster, color = cluster, fill = cluster))+
  geom_density(aes(y = ..scaled..), bw = bw, alpha = 0.2)+
  theme_light()
  print(p1)

  
  # Histogram
  p1 <- df_grouped %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(., aes(x = distances, group = cluster, color = cluster, fill = cluster, alpha = 0.6))+
  geom_histogram(binwidth = 50, alpha = 0.6, position = "identity")+
  theme_light()
  print(p1)
  
  
  #ggsave(plot=p1, filename = paste(paste(OUT_FOLDER, paste("SNVs_distribution", marker, label, "density_bandwidth", bw, "WIN", WIN, sep="_"), sep=""), ".png", sep=""), device = "png", width = 14, height = 7)
  
}

```



```{r}
for(marker in names(all_distances)){
  df <- all_distances[[marker]]
  high <- clust_all[["high"]][[marker]]
  low <- clust_all[["low"]][[marker]]
  
  dist_high <- na.omit(unlist(df[, colnames(df) %in% high, ], use.names = F))
  dist_low <-  na.omit(unlist(df[, colnames(df) %in% low, ], use.names = F))
  max_dist_length <- max(length(dist_high), length(dist_low))
  
  # How many SNVs fall in high vs. low enhancers?
  print(sprintf("Number of SNVs falling in %s enhancer region - WIN: %d", marker, WIN))
  print(sprintf("High: %d", length(dist_high)))
  print(sprintf("Low: %d", length(dist_low)))
  
  df_grouped <- data.frame("high" = c(dist_high, rep(NA, max_dist_length - length(dist_high))), 
                           "low" = c(dist_low, rep(NA, max_dist_length - length(dist_low))))
  
  perc_high <- data.frame(table(df_grouped$high))
  perc_low <- data.frame(table(df_grouped$low))
  df_grouped_perc <- full_join(perc_high, perc_low, by = "Var1")
  
  # Compute mean dividing by # of enhancers in group * 100
  size_high <- dim(enh_all[[marker]][enh_all[[marker]]$cluster %in% clust_all[["high"]][[marker]], ])[1]
  size_low <- dim(enh_all[[marker]][!enh_all[[marker]]$cluster %in% clust_all[["high"]][[marker]], ])[1]
  
  df_grouped_perc$Freq.x <- round(df_grouped_perc$Freq.x /  size_high *100,2)*100
  df_grouped_perc$Freq.y <- round(df_grouped_perc$Freq.y / size_low *100,2)*100
  
  df_grouped_perc[is.na(df_grouped_perc)] <- 0
  colnames(df_grouped_perc)[2:3] <- c("high", "low")
  
  # Plot region of +/- 500 bp
  win_to_plot <- seq(-500, 500, 1)
  df_grouped_perc_to_plot <- df_grouped_perc[df_grouped_perc$Var1 %in% win_to_plot, ]
  
  p1 <- df_grouped_perc_to_plot %>% pivot_longer(-Var1) %>%
    ggplot(aes(x=Var1, y = value, group = name, color = name))+
    geom_line()+
    scale_x_discrete(breaks = c(-500, 0, 500))+
    theme_light()
  print(p1)
    
  p2 <- df_grouped_perc_to_plot %>% pivot_longer(-Var1) %>%
    ggplot(aes(x=Var1, y = value, group = name, color = name))+
    geom_point()+
    scale_x_discrete(breaks = c(-500, 0, 500))+
    theme_light()
  print(p2)

}


```


### Distribution of STSMs breakpoint over enhancers regions ###

```{r}
### INPUT FILES
# STSMs - No start-end, but chrom_bkpt
STSMs <- read_tsv(paste(IN_FOLDER, "./genomics/pre_processed_ICGC/structural_somatic_mutation.preprocessed.tsv", sep=""))
STSMs <- STSMs[, -c(10, 12, 15, 17)]
STSMs$chrom <- paste("chr", STSMs$chrom, sep = "")
STSMs$start <- STSMs$chrom_bkpt
STSMs$end <- STSMs$chrom_bkpt

STSMs_gr <- makeGRangesFromDataFrame(STSMs, keep.extra.columns = T)

```

```{r}
label <- "all_enhancers_and_k27ac_control"
WIN <- 3000 

for(marker in names(enh_all)){
  print(marker)
  
  ### Pre/process input data
  # CtIP/GRHL enhancers
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'
  # Putative enhancers
  put_enh <- put_enhancers[sample(seq(1,dim(put_enhancers)[1]), size = dim(marker_enh)[1], replace = F), ] # sample a set of regions 
  print(paste("Number of putative enhancers sampled: ", dim(put_enh)[1]))
  
  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN
  put_enh$start <- put_enh$summit - WIN
  put_enh$end <- put_enh$summit + WIN
  
  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)
  put_enh_gr <- makeGRangesFromDataFrame(put_enh, keep.extra.columns = T)

  
  ### Find SNVs / Enhancers overlaps 
  # All CtIP & GRHL enhancers - overlap with matching SSMs 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=STSMs_gr)
  STSMs_sbj <- data.frame(STSMs_gr[subjectHits(hits_obj_enh)])
  colnames(STSMs_sbj)[1:5] <- paste(colnames(STSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), STSMs_sbj) 

  hits_obj_put_enh <- findOverlaps(query = put_enh_gr, subject = STSMs_gr)
  STSMs_sbj <- data.frame(STSMs_gr[subjectHits(hits_obj_put_enh)])
  colnames(STSMs_sbj)[1:5] <- paste(colnames(STSMs_sbj)[1:5], "sbj", sep = "_")
  put_enh_SSMs <- cbind(data.frame(put_enh_gr[queryHits(hits_obj_put_enh)]), STSMs_sbj) 
  
  
  ### Compute dist of SNVs from enhancer summit 
  dist_enh <- enh_SSMs$start_sbj - enh_SSMs$summit
  dist_put_enh <- put_enh_SSMs$start_sbj - put_enh_SSMs$summit
  max_lenght <- max(length(dist_enh), length(dist_put_enh))
  dist_all <- data.frame(dist_enh = c(dist_enh, rep(NA, max_lenght - length(dist_enh))), 
                         dist_put_enh = c(dist_put_enh, rep(NA, max_lenght - length(dist_put_enh))))
    
  
  ### Plot binned distributions
  nbins <- c(1000, 100)
  ifelse(marker == "CtIP", color <- "lightblue", color <- "pink")
  
  for(i in 1:2){
    p <- dist_all %>% ggplot(.)+
      geom_histogram(aes(x = dist_enh), fill = color, alpha = 0.7, bins = nbins[i])+
      geom_text(data = data.frame(), aes(x=Inf, y=Inf, label= marker), color=color, vjust = 1.5, hjust = 2, 
                size = 7, fontface = "bold")+
      geom_histogram(aes(x = dist_put_enh, y = -..count..), fill = "gray", alpha = 0.7, bins = nbins[i])+
      geom_text(data=data.frame(), aes(x=Inf, y=-Inf, label= "H3K27ac_CTRL"), color="gray", vjust = -1, hjust = 1.1, 
                size = 7, fontface = "bold")+
      theme_light()+
      scale_x_continuous(breaks = seq(-3000, 3000, 500))+
      labs(title = paste("Histogram of distances from peak summit - ", marker, " (Binsize = ", WIN*2/nbins, "bp)", sep =""), 
         x = "distance from summit", y = "")
    print(p)
    
    ### Save Plot
    #ggsave(paste(paste(OUT_FOLDER, paste("STSMs_distribution", marker, label, "binsize", WIN*2/nbins[i], "bp", sep="_"), sep=""), ".png", sep=""), plot = p, device = "png", width = 18, height = 7)
    
  }
  
  ## Some stats ##
  print(marker)
  print(sprintf("Number of unique bkpts overlapping: %d - %.1f%%", length(unique(enh_SSMs$sv_id_x_class)), 
                length(unique(enh_SSMs$sv_id_x_class))/dim(STSMs)[1]*100))
  print(sprintf("Number of enhancers mutated: %d - %.1f%%", length(unique(enh_SSMs$name)), 
                length(unique(enh_SSMs$name))/dim(marker_enh)[1]*100))
  
}

```


# STSMs distribution based on TYPE of STSM
```{r} 
WIN <- 3000 
bw <- 40

for(marker in names(enh_all)){
  
  ### Pre/process input data
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'

  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN

  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)
  
  
  ### Find SNVs / Enhancers overlaps 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=STSMs_gr)
  STSMs_sbj <- data.frame(STSMs_gr[subjectHits(hits_obj_enh)])
  colnames(STSMs_sbj)[1:5] <- paste(colnames(STSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), STSMs_sbj) 

  ### Compute dist of SNVs from enhancer summit 
  dist_enh <- enh_SSMs$start_sbj - enh_SSMs$summit
  max_lenght <- max(length(dist_enh))
  dist_all <- data.frame(dist_enh = dist_enh, type = enh_SSMs$variant_type)
  
  ### Plot binned distributions
  nbins <- 100
  ifelse(marker == "CtIP", color <- "lightblue", color <- "pink")
  
  p <- dist_all %>%
    ggplot(., aes(x = dist_enh, group = type, fill = type))+
    geom_histogram(position = "identity", alpha = 0.6, binwidth = bw)+
    labs(title = marker)+
    theme_light()+
    theme(legend.position = "bottom")
  print(p)
  
  #ggsave(plot = p, filename = paste(OUT_FOLDER, paste("Per_type_STSMs_distribution", marker, "bw", bw, sep="_"), ".png", sep=""), device = "png", width = 10, height = 6)

  types_plot_l <- list()
  for(i in 1:length(unique(enh_SSMs$variant_type))){
    t <- unique(enh_SSMs$variant_type)[i]
    one_type <- dist_all[dist_all$type == t, ]

    st <- str_split(t, " ", simplify = T)
    if(length(st) == 5){t <- paste(str_flatten(st[1:2], collapse = " "), "\n", 
                                   str_flatten(st[3:5], collapse = " "))}
      
    p <- one_type %>% ggplot(., aes(x = dist_enh))+
      geom_histogram(position = "identity", binwidth = bw)+
      labs(title = t)+
      theme_light()+
      theme(axis.title = element_text(size = 8))
    types_plot_l[[i]] <- p
    
  }
  
  p_grid <- cowplot::plot_grid(plotlist = types_plot_l)
  print(p_grid)
  
  #ggsave(plot = p_grid, filename = paste(OUT_FOLDER, paste("Per_type_STSMs_distribution.individual", marker, "bw", bw, sep="_"), ".png", sep=""), device = "png", width = 10, height = 6)
}

```


```{r}
for(marker in names(enh_all)){
  
  ### Pre/process input data
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'

  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN

  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)
  
  
  ### Find SNVs / Enhancers overlaps 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=STSMs_gr)
  STSMs_sbj <- data.frame(STSMs_gr[subjectHits(hits_obj_enh)])
  colnames(STSMs_sbj)[1:5] <- paste(colnames(STSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), STSMs_sbj) 

  
  for(cluster in unique(marker_enh$cluster)){
    enh_SSMs_clust <- enh_SSMs[enh_SSMs$cluster == cluster, ]

    ### Compute dist of SNVs from enhancer summit 
    dist_enh <- enh_SSMs_clust$start_sbj - enh_SSMs_clust$summit
    max_lenght <- length(dist_enh)
    dist_all <- data.frame(dist_enh = dist_enh, type = enh_SSMs_clust$variant_type)

    ### Plot binned distributions
    nbins <- 100
    ifelse(marker == "CtIP", color <- "lightblue", color <- "pink")
    
    p <- dist_all %>%
      ggplot(., aes(x = dist_enh, group = type, fill = type))+
      geom_histogram(position = "identity", alpha = 0.6, bins = nbins)+
      #geom_density(bw = 100)+
      labs(title = paste(marker, cluster, sep =" "))+
      theme_light()+
      theme(legend.position = "bottom")
    print(p)
    
    #ggsave(plot = p, filename = paste(OUT_FOLDER, "clusters/", paste("Per_type_STSMs_distribution", marker, cluster, "n_bins", nbins, sep="_"), ".png", sep=""), device = "png", width = 10, height = 6)  
  
    types_plot_l <- list()
    for(i in 1:length(unique(enh_SSMs$variant_type))){
      t <- unique(enh_SSMs$variant_type)[i]
      one_type <- dist_all[dist_all$type == t, ]
  
      st <- str_split(t, " ", simplify = T)
      if(length(st) == 5){t <- paste(str_flatten(st[1:2], collapse = " "), "\n", str_flatten(st[3:5], collapse = " "))}
        
      p <- one_type %>% ggplot(., aes(x = dist_enh))+
        geom_histogram(bins = nbins)+
        labs(title = t)+
        theme_light()+
        theme(axis.title = element_text(size = 8))
      types_plot_l[[i]] <- p
      }
  
    p_grid <- cowplot::plot_grid(plotlist = types_plot_l)
    print(p_grid)
    
    #ggsave(plot = p_grid, filename = paste(OUT_FOLDER, "clusters/", paste("Per_type_STSMs_distribution.individual", marker, cluster, "n_bins", nbins, sep="_"), ".png", sep=""), device = "png", width = 10, height = 6)
    
  }
}

```


# Position of Start/End bkpts within regions
```{r}
df_var_stats_all <- list("CtIP" = NA, "GRHL" = NA)


for(marker in names(enh_all)){
  
  ### Pre/process input data
  marker_enh <- enh_all[[marker]] 

  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN

  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)
  
  
  ### Find SNVs / Enhancers overlaps 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=STSMs_gr)
  STSMs_sbj <- data.frame(STSMs_gr[subjectHits(hits_obj_enh)])
  colnames(STSMs_sbj)[1:5] <- paste(colnames(STSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), STSMs_sbj) 
  
  # By enhancer
  df_var_stats <- data.frame(name = unique(enh_SSMs$name), 
                             n_bkpts = NA, 
                             n_single_bkpts = NA,
                             n_within_stsms = NA)
  
  for(i in 1:length(unique(enh_SSMs$name))){
    # Variants overlapping each enhancer - ACROSS DONORS
    enh <- unique(enh_SSMs$name)[i]
    SSMs_sub <- enh_SSMs[enh_SSMs$name == enh, ]
  
    df_var_stats[df_var_stats$name == enh, "n_bkpts"] <- dim(SSMs_sub)[1]
    
    # For each enhancer, are there SVs with 2 bkpts within the enhancer?
    n_unique_vars <- table(SSMs_sub$sv_id)
    
    # For each var, only 1 instance. Only start or end bkpt falls within enhancer
    if(any(n_unique_vars > 1) == F){
      df_var_stats[df_var_stats$name == enh, "n_single_bkpts"] <- dim(SSMs_sub)[1]
      df_var_stats[df_var_stats$name == enh, "n_within_stsms"] <- 0
    } else if(any(n_unique_vars > 1) == T){
      df_var_stats[df_var_stats$name == enh, "n_single_bkpts"] <- sum(n_unique_vars == 1)
      
      n_single <- 0
      n_within <- 0
      
      # Which stsms have more than 1 bkpt
      SSMs_sub <- SSMs_sub %>% filter(sv_id %in% names(n_unique_vars[n_unique_vars > 1]))
      non_unique_vars <- unique(SSMs_sub$sv_id)
      for(i in 1:length(non_unique_vars)){
        one_var_df <- SSMs_sub[SSMs_sub$sv_id == non_unique_vars[i], ]
        # How many unique sv_id_x_class (from/to) x icgc_donor_id 
        df_stat <- one_var_df %>% group_by(icgc_donor_id) %>% summarise(n_sv_id_x_class = length(unique(sv_id_x_class)))
        n_within <- n_within + sum(df_stat$n_sv_id_x_class == 2)
        n_single <- n_single + sum(df_stat$n_sv_id_x_class == 1)
      }
      
      df_var_stats[df_var_stats$name == enh, "n_within_stsms"] <- n_within
      df_var_stats[df_var_stats$name == enh, "n_single_bkpts"] <- df_var_stats[df_var_stats$name == enh, "n_single_bkpts"]+n_single
    }
  }
  
  df_var_stats_all[[marker]] <- df_var_stats
}


for(marker in names(enh_all)){
  var_stats_marker <- df_var_stats_all[[marker]]
  print(apply(var_stats_marker[,-1], MARGIN = 2, FUN = sum))

  print(sprintf("Percentage of single bkpts: %.0f%%", 
    apply(var_stats_marker[,-1], MARGIN = 2, FUN = sum)[2] / apply(var_stats_marker[,-1], MARGIN = 2, FUN = sum)[1]*100))
  print(sprintf("Percentage of within bkpts: %.0f%%", 
    apply(var_stats_marker[,-1], MARGIN = 2, FUN = sum)[3] / apply(var_stats_marker[,-1], MARGIN = 2, FUN = sum)[1]*100))
}

aa <- left_join(var_stats_marker[var_stats_marker$n_within_stsms != 0, ], enh_SSMs[, c("name", "cluster")], by = "name")
aa <- aa[!duplicated(aa), ]
table(aa$cluster)

```

# Deletions spanning the full enhancer
```{r}
deletions <- STSMs %>% filter(variant_type == "deletion")
length(unique(deletions$sv_id))
dim(deletions)[1]

deletions_from <- deletions %>% filter(class == "from") %>%
  select(-c(chrom_bkpt, end, range, class))
deletions_to <- deletions %>% filter(class == "to") %>%
  select(sv_id, start)
deletions_coords <- left_join(deletions_from, deletions_to, by = "sv_id")
colnames(deletions_coords)[12:13] <- c("start", "end")

# Length of deletions
data.frame(x = deletions_coords$end - deletions_coords$start) %>%
  ggplot(., aes(x = x))+
  geom_histogram()+
  scale_x_continuous(limits = c(0, 5.0e+07))
summary(deletions_coords$end - deletions_coords$start)

deletions_gr <- makeGRangesFromDataFrame(deletions_coords, keep.extra.columns = T)


for(marker in names(enh_all)){
  marker_enh <- enh_all[[marker]] 
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN

  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)
  
  # if type is within, the query interval must be wholly contained within the subject interval. 
  hits_obj <- findOverlaps(query = enh_gr, subject = deletions_gr, 
             type = "within", minoverlap = 1)
  del_sbj <- data.frame(deletions_gr[subjectHits(hits_obj)])
  colnames(del_sbj)[1:5] <- paste(colnames(del_sbj)[1:5], "sbj", sep = "_")
  enh_dels <- cbind(data.frame(enh_gr[queryHits(hits_obj)]), del_sbj) 
  
  sum(enh_dels$end_sbj - enh_dels$end < 0) # should be 0 - STSM end always after enhancer end
  sum(enh_dels$start_sbj - enh_dels$start > 0) # should be 0 - STSM start always before enhancer start
  
  cat("\n")
  print(marker)
  print(sprintf("Number of deletions spanning a full enhancer: %.0f - %.1f%%", length(unique(enh_dels$sv_id)), 
                length(unique(enh_dels$sv_id))/dim(deletions_coords)[1]*100))  
  print(sprintf("Number of enhancers fully deleted: %.0f - %.1f%%", length(unique(enh_dels$name)), 
                length(unique(enh_dels$name))/dim(marker_enh)[1]*100))
  print(sprintf("Number of donors hit: %.0f",length(unique(enh_dels$icgc_donor_id))))
  
  # Number of enhancers removed by each deletion
  n_del_enh <- enh_dels %>% group_by(sv_id) %>% 
    dplyr::summarise(n_enh_deleted = length(unique(name))) %>%
    arrange(-n_enh_deleted)

  print("Number of enhancers removed by each deletion:")
  print(summary(n_del_enh$n_enh_deleted))
  print("Number of deletions removing only one enhancer:")
  print(dim(n_del_enh[n_del_enh$n_enh_deleted == 1, ])[1])
  
  
  # Number of donors harboring an 'enhancer-specific' deletion
  enh_spec_del <- n_del_enh[n_del_enh$n_enh_deleted == 1, ]$sv_id
  n_enh_spec_dels <- enh_dels %>% filter(sv_id %in% enh_spec_del) %>% 
    group_by(icgc_donor_id) %>% 
    dplyr::summarise(n_enh_spec_dels = length(unique(sv_id)))
  n_enh_spec_dels <- rbind(n_enh_spec_dels,
    data.frame("icgc_donor_id" = unique(STSMs$icgc_donor_id[!STSMs$icgc_donor_id %in% n_enh_spec_dels$icgc_donor_id]),
             "n_enh_spec_dels" = 0))
  
  print("Number of 'enhancer-specific' dels x donor:")
  print(summary(n_enh_spec_dels$n_enh_spec_dels))
  
}

```



## STSMs distribution based on clusters ##
```{r}
save_plots <- FALSE

label <- "clustered_enhancers"
dist_clustered_hist <- list("CtIP" = setNames(replicate(length(unique(enh_all[["CtIP"]]$cluster)), list()),
                                               unique(enh_all[["CtIP"]]$cluster)), 
                             "GRHL" = setNames(replicate(length(unique(enh_all[["GRHL"]]$cluster)), list()),
                                               unique(enh_all[["GRHL"]]$cluster)))

for(marker in names(enh_all)){

  ### Pre/process input data
  # CtIP/GRHL enhancers
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'
  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN
  
  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)

  
  ### Find SNVs / Enhancers overlaps 
  # All CtIP & GRHL enhancers - overlap with matching SSMs 
  ### Find SNVs / Enhancers overlaps 
  # All CtIP & GRHL enhancers - overlap with matching SSMs 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=STSMs_gr)
  STSMs_sbj <- data.frame(STSMs_gr[subjectHits(hits_obj_enh)])
  colnames(STSMs_sbj)[1:5] <- paste(colnames(STSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), STSMs_sbj) 


  for(cluster in unique(marker_enh$cluster)){
    enh_SSMs_clust <- enh_SSMs[enh_SSMs$cluster == cluster, ]

    ### Compute dist of SNVs from enhancer summit 
    dist_enh <- enh_SSMs_clust$start_sbj - enh_SSMs_clust$summit
    max_lenght <- max(length(dist_enh), length(dist_put_enh))
    dist_all <- data.frame(dist_enh = c(dist_enh, rep(NA, max_lenght - length(dist_enh))))
    
    ### Plot binned distributions
    nbins <- c(100)
    ifelse(marker == "CtIP", color <- "lightblue", color <- "pink")
    title <- paste("Histogram of distances from peak summit - ", marker, " (Binsize = ", WIN*2/nbins, "bp)", sep ="")
    
    p <- dist_all %>% ggplot(.)+
      geom_histogram(aes(x = dist_enh), fill = color, alpha = 0.7, bins = nbins[1])+
      geom_text(data = data.frame(), aes(x=Inf, y=Inf, label= marker), color=color, vjust = 1.5, hjust = 2, 
                size = 7, fontface = "bold")+
      theme_light()+
      scale_x_continuous(breaks = seq(-3000, 3000, 500))+
      labs(title = NULL, 
           subtitle = cluster,
         x = "distance from summit", y = "")
    dist_clustered_hist[[marker]][[cluster]] <- p
    print(p)
    
  }
  
  # Save grid of plots 
  if(save_plots == T){
    title <- cowplot::ggdraw() + cowplot::draw_label(title, fontface = "bold")
    cowplot::plot_grid(title, plotlist = dist_clustered_hist[[marker]], ncol = 1) %>%
    ggsave(., filename = paste(paste(OUT_FOLDER, 
                       paste("STSMs_distribution", marker, label, "binsize", WIN*2/nbins[1], "bp", sep="_"), sep=""), ".png", sep=""), 
        device = "png", width = 7, height = 18)

        
  }
}

```




# Density plot
## STSMs distribution based on clusters ##
```{r}
save_hist <- FALSE
save_density <- FALSE

label <- "clustered_enhancers"
WIN <- 3000
dist_clustered_hist <- list("CtIP" = setNames(replicate(length(unique(enh_all[["CtIP"]]$cluster)), list()),
                                               unique(enh_all[["CtIP"]]$cluster)), 
                             "GRHL" = setNames(replicate(length(unique(enh_all[["GRHL"]]$cluster)), list()),
                                               unique(enh_all[["GRHL"]]$cluster)))
dist_clustered_density <- list("CtIP" = setNames(replicate(length(unique(enh_all[["CtIP"]]$cluster)), list()),
                                               unique(enh_all[["CtIP"]]$cluster)), 
                             "GRHL" = setNames(replicate(length(unique(enh_all[["GRHL"]]$cluster)), list()),
                                               unique(enh_all[["GRHL"]]$cluster)))

# Save all distances to be able to plot them together
all_distances <- list("CtIP" = NA, "GRHL" = NA)

for(marker in names(enh_all)){

  ### Pre/process input data
  # CtIP/GRHL enhancers
  marker_enh <- enh_all[[marker]] # select enhancer with 'marker'
  ### Extend regions of 'WIN' from summit
  marker_enh$start <- marker_enh$summit - WIN
  marker_enh$end <- marker_enh$summit + WIN
  
  enh_gr <- makeGRangesFromDataFrame(marker_enh, keep.extra.columns = T)

  
  ### Find SNVs / Enhancers overlaps 
  # All CtIP & GRHL enhancers - overlap with matching SSMs 
  hits_obj_enh <- findOverlaps(query=enh_gr, subject=STSMs_gr)
  STSMs_sbj <- data.frame(STSMs_gr[subjectHits(hits_obj_enh)])
  colnames(STSMs_sbj)[1:5] <- paste(colnames(STSMs_sbj)[1:5], "sbj", sep = "_")
  enh_SSMs <- cbind(data.frame(enh_gr[queryHits(hits_obj_enh)]), STSMs_sbj) 

  # Save all distances to be able to plot them together
  all_distances_marker <- data.frame(matrix(nrow = max(table(enh_SSMs$cluster)), ncol = length(unique(enh_SSMs$cluster))))
  colnames(all_distances_marker) <- unique(enh_SSMs$cluster)
  for(cluster in unique(marker_enh$cluster)){
    enh_SSMs_clust <- enh_SSMs[enh_SSMs$cluster == cluster, ]

    ### Compute dist of SNVs from enhancer summit 
    dist_enh <- enh_SSMs_clust$start_sbj - enh_SSMs_clust$summit
    max_lenght <- max(length(dist_enh))
    dist_all <- data.frame(dist_enh = c(dist_enh, rep(NA, max_lenght - length(dist_enh))))
    
    all_distances_marker[, cluster] <- c(dist_all[, "dist_enh"], rep(NA, dim(all_distances_marker)[1] - 
                                                                       length(dist_all[, "dist_enh"])))
    
    ### Plot binned distributions
    nbins <- c(100)
    ifelse(marker == "CtIP", color <- "lightblue", color <- "pink")
    title <- paste("Histogram of distances from peak summit - ", marker, " (Binsize = ", WIN*2/nbins, "bp)", sep ="")
    
    p <- dist_all %>% ggplot(.)+
      geom_histogram(aes(x = dist_enh), fill = color, alpha = 0.7, bins = nbins[1])+
      geom_text(data = data.frame(), aes(x=Inf, y=Inf, label= marker), color=color, vjust = 1.5, hjust = 2, 
                size = 7, fontface = "bold")+
      theme_light()+
      scale_x_continuous(breaks = seq(-3000, 3000, 500))+
      labs(title = NULL, 
           subtitle = cluster,
         x = "distance from summit", y = "")
    dist_clustered_hist[[marker]][[cluster]] <- p
    #print(p)
    
    title <- paste("Density of distances from peak summit - ", marker, sep ="")
    bw = 100
    dens_p <- dist_all %>% ggplot(.)+
      geom_density(aes(x = dist_enh), fill = color, color = color, alpha = 0.7, bw = bw)+
      geom_text(data = data.frame(), aes(x=Inf, y=Inf, label= marker), color=color, vjust = 1.5, hjust = 2, 
                size = 7, fontface = "bold")+
      theme_light()+
      scale_x_continuous(breaks = seq(-3000, 3000, 500))+
      labs(title = NULL, 
           subtitle = cluster,
         x = "distance from summit", y = "")
    dist_clustered_density[[marker]][[cluster]] <- dens_p
    
    print(dens_p)
    
  }
  
  all_distances[[marker]] <- all_distances_marker
  
  # Save grid of plots 
  if(save_hist == T){
    title <- cowplot::ggdraw() + cowplot::draw_label(title, fontface = "bold")
    cowplot::plot_grid(title, plotlist = dist_clustered_hist[[marker]], ncol = 1) %>%
    ggsave(., filename = paste(paste(OUT_FOLDER, 
                       paste("STSMs_distribution", marker, label, "binsize", WIN*2/nbins[1], "bp", sep="_"), sep=""), ".png", sep=""), 
        device = "png", width = 7, height = 20)
  }
  
  if(save_density == T){
    title <- cowplot::ggdraw() + cowplot::draw_label(title, fontface = "bold")
    cowplot::plot_grid(title, plotlist = dist_clustered_density[[marker]], ncol = 1) %>%
    ggsave(., filename = paste(paste(OUT_FOLDER, 
          paste("STSMs_distribution", marker, label, "density_bandwidth", bw, "WIN", WIN, sep="_"), sep=""), ".png", sep=""),           device = "png", width = 7, height = 20)
  }
  
}


```


```{r}

for(marker in names(all_distances)){
  df <- all_distances[[marker]] 
  p <-  df %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(data=., aes(x=distances, group=cluster, color = cluster)) +
    geom_density(bw = 200, alpha = 0.1) +
    theme_light()
  print(p)
}

# Grouping distances based on high-low clusters
for(marker in names(all_distances)){
  df <- all_distances[[marker]]
  high <- clust_high[[marker]]
  low <- clust_low[[marker]]
  
  dist_high <- na.omit(unlist(df[, colnames(df) %in% high, ], use.names = F))
  dist_low <-  na.omit(unlist(df[, colnames(df) %in% low, ], use.names = F))
  max_dist_length <- max(length(dist_high), length(dist_low))
  
  df_grouped <- data.frame("high" = c(dist_high, rep(NA, max_dist_length - length(dist_high))), 
                           "low" = c(dist_low, rep(NA, max_dist_length - length(dist_low))))
  
  p1 <- df_grouped %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(., aes(x = distances, group = cluster, color = cluster, fill = cluster))+
  geom_density(bw = 100, alpha = 0.2)+
  theme_light()
  print(p1)
  
  p2 <- df_grouped %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(., aes(x = distances, group = cluster, color = cluster))+
  geom_density(bw = 100, alpha = 0.2, size = 1.5)+
  theme_light()
  print(p2)
  
  p3 <- df_grouped %>% pivot_longer(everything(), names_to = "cluster", values_to = "distances") %>%
  ggplot(., aes(x = distances, group = cluster, color = cluster))+
  geom_histogram(bw = 100, alpha = 0.2, position = "identity")+
  theme_light()
  print(p3)
}



```

